documentation/projects/proposals/search_relevancy_sandbox/20230518-implementation_plan_staging_index_rapid_iteration.md

```diff
@@ -0,0 +1,357 @@
+
+Once the implementation is complete, a maintainer should be able to run this DAG
+to create a new index on either elasticsearch environment with an altered
+configuration. The maintainer should only need to specify the pieces of the
+configuration that are different from the default configuration, although they
+can supply a full configuration if they wish. The DAG will report when it is
+complete and what the name of the new index is.
+
+## Dependencies
+
+The following are prerequisites for the DAG:
+
+- Installation of the
+  [`elasticsearch` Airflow provider package](https://airflow.apache.org/docs/apache-airflow-providers-elasticsearch/stable/)
+- Airflow Connections for each Elasticsearch cluster
+- Sharing of the
+  [`es_mappings.py` configuration](https://github.com/WordPress/openverse/blob/0a5f4ab2ce5d80a48bd1c57d2a2dbcca14fcbedc/ingestion_server/ingestion_server/es_mapping.py)
+
+### Elasticsearch provider
+
+The
+[`elasticsearch` provider](https://airflow.apache.org/docs/apache-airflow-providers-elasticsearch/stable/)
+will need to be added to the list of dependencies for our catalog Docker image.
+This will be done first and deployed on its own, as the connections will be
+needed for the rest of the work.
+
+The
+[production requirements](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/catalog/requirements_prod.txt#L5)
+will need to be changed from `apache-airflow[amazon,postgres,http]` to
+`apache-airflow[amazon,postgres,http,elasticsearch]`.
+
+### Airflow connections
+
+Once the provider is installed,
+[Airflow Connections](https://airflow.apache.org/docs/apache-airflow-providers-elasticsearch/stable/connections/elasticsearch.html)
+will need to be created for each Elasticsearch cluster. The connections will be
+named `elasticsearch_production` and `elasticsearch_staging` and reference each
+respective environment. These will be created by hand in the Airflow UI. For
+local development, defaults will be added to the
+[`env.template` file](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/catalog/env.template).
+
+### Sharing `es_mappings.py`
+
+The
+[`es_mappings.py` configuration](https://github.com/WordPress/openverse/blob/0a5f4ab2ce5d80a48bd1c57d2a2dbcca14fcbedc/ingestion_server/ingestion_server/es_mapping.py)
+in the `ingestion_server` module defines, in python, the default index
+configuration we use for all media types. This configuration does not require
+any additional Python dependencies, but does not exist in a form that can be
+imported by the DAG. To make this configuration available to the DAG, we will
+add the `es_mapping.py` file to the
+[`sync.yml` workflow configuration](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/.github/sync.yml)
+to keep the file up-to-date across both services. The destination for this sync
+could be `catalog/dags/common/elasticsearch/es_mapping.py`.
```

Comment:
> Would we be able to avoid this by using the [get index settings API](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-get-settings.html) to retrieve the settings on the current origin index for the media type being copied and base the new settings on that instead?

That would be very cool! And avoids the concerns described [here](https://github.com/WordPress/openverse/pull/2133/files#diff-344cb6ad9993d7e365933d71faa76a18fbbb062a333a5ae4f73b9dee861b46d8R281) as well.

Reply:
Great idea, I'll look into this!

Reply:
Note for myself here - we can definitely use the existing API rather than a repo sync, but it will require some massaging of the response data. For instance, Current form of the `index_setting` function spits out:

```json
{
  "settings": {
    "index": {
      "number_of_shards": 18,
      "number_of_replicas": 0,
      "refresh_interval": "-1"
    },
    "analysis": {...},
  },
  "mapping": {...}
}
```

However, the response returned from `GET /<index-or-alias>/`:

```json
{
  "<index-name>": {
    "settings": {
      "index": {
        "number_of_shards": "18",
        "number_of_replicas": 0,
        "refresh_interval": "-1",
        ...
        "analysis": {
          ...
        }
      }
    },
    "mappings": {...}
  }
}
```

I'll add this to the IP, just didn't want to forget it!